apps/extension/src/contents/relay.ts

```diff
@@ -11,30 +11,25 @@ export const config: PlasmoCSConfig = {
   run_at: "document_start"
 }
 
-const ports: Record<PortName, Port | undefined> = {
-  [PortName.Completion]: undefined,
-  [PortName.Model]: undefined,
-  [PortName.Events]: undefined,
-  [PortName.Permission]: undefined
+const portStates: Record<
+  PortName,
+  { port?: Port; listener?: (message: PortEvent[PortName], port: Port) => void }
+> = {
+  [PortName.Completion]: {},
+  [PortName.Model]: {},
+  [PortName.Events]: {},
+  [PortName.Permission]: {}
 }
 
-function connectWithRetry(portName: PortName): Port {
+function initPortState(portName: PortName) {
   log(`Connecting to ${portName} port`)
+  const portState = portStates[portName]
   const port = Extension.connectToBackground(portName, () =>
-    connectWithRetry(portName)
+    initPortState(portName)
```

Comment:
I think the disconnect callback is the best place to call the remove listener, right before the initPortState :-?... 

Something like this maybe:

```ts
Extension.removePortListener(portState.listener, portState.port);
delete portState.port;
delete portState.listener;
```

Since the JS's GC might take its time

Reply:
done!